plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs

```diff
@@ -149,98 +150,172 @@ At this point in the pipeline the type of the denotation of a builtin is assumed
--- the optimization.
+-- We could've computed the runtime denotations ('toImmediateF' and 'toDeferredF') from the
+-- 'TypeScheme' and the denotation of the builtin, but not statically (due to unfolding not working
+-- for recursive functions and 'TypeScheme' being recursive, i.e. requiring the conversion function
+-- to be recursive), and so it would cause us to retain a lot of evaluation-irrelevant stuff in the
+-- constructors of 'BuiltinRuntime', which has to make evaluation slower (we didn't check) and
+-- certainly makes the generated Core much harder to read. Technically speaking, we could get
+-- a 'RuntimeScheme' from the 'TypeScheme' and the denotation statically if we changed the
+-- definition of 'TypeScheme' and made it a singleton, but then the conversion function would have
+-- to become a class anyway and we'd just replicate what we have here, except in a much more
+-- complicated way.
 class KnownMonotype val args res where
     knownMonotype :: TypeScheme val args res
-    knownMonoruntime :: RuntimeScheme (Length args)
 
     -- | Convert the denotation of a builtin to its runtime counterpart with immediate unlifting.
-    toImmediateF :: FoldArgs args res -> ToRuntimeDenotationType val (Length args)
+    -- We use a tuple rather than two arguments for symmetry with 'toPolyDeferredF'. It all gets
+    -- inlined anyway.
+    toMonoImmediateF
+        :: (FoldArgs args res, FoldArgs args ExBudget)
+        -> BuiltinRuntime val
 
     -- | Convert the denotation of a builtin to its runtime counterpart with deferred unlifting.
     -- The argument is in 'ReadKnownM', because that's what deferred unlifting amounts to:
     -- passing the action returning the builtin application around until full saturation, which is
     -- when the action actually gets run.
-    toDeferredF :: ReadKnownM (FoldArgs args res) -> ToRuntimeDenotationType val (Length args)
+    toMonoDeferredF
+        :: ReadKnownM (FoldArgs args res, FoldArgs args ExBudget)
+        -> BuiltinRuntime val
 
 -- | Once we've run out of term-level arguments, we return a
 -- 'TypeSchemeResult'/'RuntimeSchemeResult'.
 instance (Typeable res, KnownTypeAst (UniOf val) res, MakeKnown val res) =>
             KnownMonotype val '[] res where
     knownMonotype = TypeSchemeResult
-    knownMonoruntime = RuntimeSchemeResult
 
-    toImmediateF = makeKnown
-    {-# INLINE toImmediateF #-}
+    toMonoImmediateF (x, cost) = BuiltinResult cost $ makeKnown x
+    {-# INLINE toMonoImmediateF #-}
 
     -- For deferred unlifting we need to lift the 'ReadKnownM' action into 'MakeKnownM',
     -- hence 'liftReadKnownM'.
-    toDeferredF getRes = liftReadKnownM getRes >>= makeKnown
-    {-# INLINE toDeferredF #-}
+    toMonoDeferredF =
+        either
+            -- Unlifting has failed and we don't care about costing at this point, since we're about
+            -- to terminate evaluation anyway, hence we put 'mempty' as the cost of the operation.
```

Comment:
> I guess conceivably if we specified the cost result on failure this could make it difficult to specify, but I don't think we need to do that.

What does it mean to compute the cost of an impossible operation? Sure you can extract an `ExMemory` from the term provided as an argument to a builtin, but if the builtin expects a `Bool` and was given an `Integer` or even a function instead, what would it even mean to call the costing function over the `ExMemory` of this argument?

Reply:
Ah right, we don't incur any additional costs from unlifting, so it is indeed meaningless.